numpy/_core/src/multiarray/lowlevel_strided_loops.c.src

```diff
@@ -1460,6 +1460,9 @@ mapiter_trivial_@name@(
     char *base_ptr, *ind_ptr, *result_ptr;
     npy_intp self_stride, ind_stride, result_stride;
     npy_intp fancy_dim = PyArray_DIM(self, 0);
+    if (PyArray_DATA(self) == PyArray_DATA(result)) {
```

Comment:
We should use the may share memory functionality here. Also would be good to do this at the call site (mapping.c), I think and make sure to decref the array again.

Reply:
I revised the code to use `solve_may_share_memory` one level up in the control flow, in `mapping.c`.

> make sure to decref the array again

I couldn't replicate the need for that, which array am I increasing the refcount for? Don't we proceed through a deep copy here?

I think this only covers a very well-behaved/specialized fancy indexing codepath based on the code comments, though that's also true of the regression test, 1-D contiguous integer data, etc.

Reply:
Right, I think it nicer, although I also think it should be even further up maybe.

Some notes:
* We ignore overlap in the `indices` here.  We actually have a path `index_has_memory_overlap` which is used indirectly by the code from `ufunc.at` and does all the checks necessary *including* for index arrays.
* I see no reason why you cannot move the check to before the fast-path `if`?  In fact `tmp_arr` gets prepared just above, which seems like a perfect opportunity to do that check.
  (I suppose in theory, overlap is possible even if a fresh array is created with `FromAny`, so could do it after the `if` rather than before.)


One thing that might be nice, is that I am pretty sure we already guard for the other copy paths against this, e.g.:
```
arr = np.arange(100)
arr[10:] = arr[:-10]
```
would be nice to check we have tests and add one together with the others if not (so we can proof we always make a defensive copy in indexing paths).